File: src/main/scala/org/muieer/scala/leetcode/LeetCode110.scala
package org.muieer.scala.leetcode

object LeetCode110 {

  def main(args: Array[String]): Unit = {
    println(isBalanced(new TreeNode(1)))
  }

  var balanced: Boolean = true

  def isBalanced(root: TreeNode): Boolean = {
    if (root == null) { true }
    else {
      fun(root)
      balanced
    }
  }

  def fun(root: TreeNode): Unit = {
    if (root == null) {}
    else {
      if (balanced) {
        balanced = Math.abs(compute(root.left) - compute(root.right)) <= 1
      }
      fun(root.left)
      fun(root.right)
    }
  }

  def compute(root: TreeNode): Int = {
    if (root == null) { 0 }
    else {
      Math.max(compute(root.left), compute(root.right)) + 1
    }
  }

}

File: src/main/java/org/muieer/java/leetcode/LeetCode110.java
package org.muieer.java.leetcode;

public class LeetCode110 {

    boolean balanced = true;

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        else {
            fun(root);
            return balanced;
        }
    }

    public void fun(TreeNode root) {
        if (root == null) {}
        else {
            if (balanced) {
                balanced = Math.abs(compute(root.left) - compute(root.right)) <= 1;
            }
            if (!balanced) {
                return;
            }
            fun(root.left);
            fun(root.right);
        }
    }

    public int compute(TreeNode root) {
        if (root == null) { return 0; }
        else {
            return Math.max(compute(root.left), compute(root.right)) + 1;
        }
    }
}
